packages/material-ui-styles/src/makeStyles.js

```diff
@@ -199,55 +191,52 @@ function makeStyles(stylesOrCreator, options = {}) {
       ...stylesOptions2,
     };
 
-    const { current: state } = React.useRef({});
+    const { current: instance } = React.useRef({
+      // previous: null,
+      // current: null,
+    });
+    const firstRender = React.useRef();
 
+    // ⚠️ You may rely on React.useMemo as a performance optimization, not as a semantic guarantee.
+    // https://reactjs.org/docs/hooks-reference.html#usememo
+    //
     // Execute synchronously every time the theme changes.
     React.useMemo(() => {
-      attach({
+      instance.current = {
         name,
-        props,
-        state,
+        state: {},
         stylesCreator,
         stylesOptions,
         theme,
-      });
+      };
+      attach(instance.current, props);
+      firstRender.current = true;
+
+      if (instance.previous) {
+        const previous = instance.previous;
+        setTimeout(() => {
```

Comment:
That will almost certainly cause nightmares in concurrent react. Why isn't this a cleanup in an effect?

Reply:
We can move this in a useEffect teardown.

Reply:
What do you have in mind as a possible issue with concurrent React?